Source/WebCore/html/HTMLElement.cpp

```diff
@@ -130,7 +130,9 @@ String HTMLElement::nodeName() const
 static inline CSSValueID unicodeBidiAttributeForDirAuto(HTMLElement& element)
 {
     ASSERT(!element.hasTagName(bdoTag));
-    if (element.hasTagName(preTag) || element.hasTagName(textareaTag))
+    ASSERT(!element.hasTagName(preTag));
+    ASSERT(!element.hasTagName(textareaTag));
+    if (RefPtr input = dynamicDowncast<HTMLInputElement>(element); input && (input->isTelephoneField() || input->isEmailField() || input->isSearchField() || input->isURLField()))
```

Comment:
I think you can check `input->isTextType()` rather than each type individually.

Reply:
I didn't because it also includes `text` field and `password` field, where we don't want to apply `unicode-bidi` rule.

https://github.com/WebKit/WebKit/blob/be51810121fcef51dde3e166d85055d8886f01e1/Source/WebCore/html/InputType.h#L100

Reply:
We don't? That doesn't seem right. Search and Text should be handled the same way.

> When an input element's dir attribute is in the auto state and its type attribute is in the Text state, then the user agent is expected to act as if it had a user-agent-level style sheet rule setting the 'unicode-bidi' property to 'plaintext'.

Not sure about password, but it also seems odd for it to be the odd one out...

Reply:
I am trying to do this part:

```
input[dir=auto i]:is([type=search i], [type=tel i], [type=url i],
[type=email i]){
  unicode-bidi: plaintext;
}
```

from web specification here, where if the input types are following and these were the tests, we are failing.

Reply:
Either, we delete `unicodeBidiAttributeForDirAuto` and apply them in `html.css` or we do above change. I am up for it as qwll, although this suggested, I haven't tested. :-)